static/tutorial/exception-safety.md

```diff
@@ -0,0 +1,341 @@
+on success will both handles remain open. This is the behavior desired
+of a well written allocation function.
+
+The other thing to point out is that _we don't mask exceptions_. It is
+_not_ the responsibility of the allocation function to mask
+asynchronous exceptions, it can rely on the fact that the caller has
+done so already. The logic behind this is: there is no way for the
+allocation function to mask exceptions outside of itself, and doing so
+is necessary for proper exception safety. Since the allocator has no
+power to do this itself, it demands masking as a precondition to being
+called.
+
+## Interruptible actions
+
+Before we can jump into cleanup actions, we need to take a slight
+detour and discuss interruptible actions. These are
+[described in the Control.Exception docs](https://www.stackage.org/haddock/nightly-2016-07-17/base-4.9.0.0/Control-Exception.html#g:13). The
+basic idea is that these are actions which, even when async exceptions
+are masked, can throw an asynchronous exception. These doesn't violate
+any of our comments above, since the exceptions are being thrown from
+an `IO` action. However, the exceptions are being generated as a kill
+signal from outside of our current subroutine, and therefore should be
+treated as asynchronous/unrecoverable exceptions.
+
+It's important to understand that the purpose of allowing for such a
+concept as an interruptible action is to *avoid deadlocks*. However,
+the tradeoff is that interruptible actions make it more difficult to
+write exception safe code, since they have allowed yet another
+"wormhole" through which exceptions can tunnel into our code. The
+simplest approach to dealing with this is: assume all `IO` actions may
+throw an exception. In practice, there is a subset of non-blocking
+`IO` actions which are known not to be interruptible, and are listed
+in the `Control.Exception` documentation linked to above.
+
+Even these interrupting exceptions can be masked as well, using the
+`uninterruptibleMask` function. Inside an `uninterrupibleMask`, it's
+still entirely possible for a synchronous exception to be generated
+(via `throwIO` or equivalent), but no asynchronous exceptions will be
+thrown. This is important as we dive into cleanup functions.
+
+## Cleanup functions
+
+Similar to allocation functions, cleanup functions need guarantees
+that an asynchronous exception will not be thrown before it starts
+executing, and therefore it's necessary for a function like `bracket`
+or `withFile` to mask async exceptions. However, it's less clear that
+uninterruptible masking is the right thing. There is a long discussion
+of this
+[on the safe-exceptions issue tracker](https://github.com/fpco/safe-exceptions/issues/3). Ultimately,
+the decision from that disucssion is that it is better to err on the
+side of caution and use uninterruptible masking, at the possible risk
+of introducing delays or deadlocks. (See the section on `hClose` and
+flushing for more details.)
```

Comment:
I understand that you don't want to delve into details, but I'd try hard to give at least an idea on why allocators are different from cleanups, and thus require different guarantees. I feel it'd help people retain this difference in requirements/assumptions much better.